src/main/java/io/cryostat/discovery/AbstractNodeTypeAdapter.java

```diff
@@ -190,10 +197,18 @@ public AbstractNode read(JsonReader reader) throws IOException {
                         }
                     }
                     reader.endObject();
-                    target = new ServiceRef(connectUrl, alias);
+
+                    // To prevent Gson recursion, we need to manually construct the ServiceRef with
+                    // a null jvmId and use that ServiceRef object to get the jvmId with correct
+                    // credentials without calling platformClient.listDiscoverableServices() which
+                    // would recurse through this adapter again to getJvmId
+                    target = new ServiceRef(null, connectUrl, alias);
```

Comment:
Hmm. Generally these deserializers and adapters just read the JSON content and build the right object instance directly out of that. In this case I would expect that either the jvmId is ignored on deserialization and we lazy-initialize it later on as needed (not during JSON deserialization), or the jvmId is taken at face value if it is in the serialized JSON form and we just assume that's correct. I haven't thought too deep into this but I would tend toward the lazy-init solution. This way we aren't just blindly trusting that the hash ID supplied by clients is correct, but we also aren't doing a heavyweight network operation during a simple JSON deserialization.

I'm imagining something like a method on the JvmIdHelper:

```java
public ServiceRef resolveId(ServiceRef sr) {
  if (sr.getJvmId() != null) return sr;
  // use tcm to compute ID
  return new ServiceRef(id, ... );
}
```

This might be annoying to use though because almost anywhere that has a `ServiceRef` would need to then also have the `JvmIdHelper` and try to resolve the IDs at every public method boundary.

Or maybe it's enough to handle that resolution inside the `DiscoveryStorage#update` method? Everything flows into the storage model through there, and everything else that queries it gets data flowing out from there.

Reply:
Scrolling through the rest of the diff, I do see that the `JvmIdHelper` has been added to the `PlatformClient` implementations that hook into the `BuiltInDiscovery`. I think this can be refactored as I was explaining above and all that shared logic around resolving the hash ID for a ServiceRef can be pulled into `DiscoveryStorage`, rather than being handled in each platform client plus the custom targets API handler - they're all pushing that data into the same storage anyway.

If the hash ID resolution fails then the storage can examine the failure reason:
1. If it's a JMX SSL failure, just log a warning and complete the plugin update with a null ID for that ServiceRef
2. If it's a JMX Auth failure, same
3. If it's a network exception like the hostname specified in the connectUrl could not be resolved, or the connection timed out, etc. then fail the update. Either by removing that single element that could not be reached, or probably just by throwing an exception and letting that bubble up to whatever plugin tried to publish this update.